src/Bicep.Core/Emit/TemplateEmitter.cs

```diff
@@ -70,7 +84,8 @@ private EmitResult EmitOrFail(Action write)
             {
                 return new EmitResult(EmitStatus.Failed, diagnostics);
             }
-
+            
+            this.CalculateTemplateHash();
```

Comment:
>CalculateTemplateHash [](start = 17, length = 21)

Can we avoid the side-effect on this function by adding passing the hash result via the `write` action? (Functions with side effects are harder to maintain, so it's better to just avoid them as much as possible.)